"use strict";(self.webpackChunk=self.webpackChunk||[]).push([[91691],{603905:(e,t,n)=>{n.d(t,{Zo:()=>p,kt:()=>m});var r=n(667294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function l(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var s=r.createContext({}),d=function(e){var t=r.useContext(s),n=t;return e&&(n="function"==typeof e?e(t):i(i({},t),e)),n},p=function(e){var t=d(e.components);return r.createElement(s.Provider,{value:t},e.children)},c={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},u=r.forwardRef((function(e,t){var n=e.components,a=e.mdxType,o=e.originalType,s=e.parentName,p=l(e,["components","mdxType","originalType","parentName"]),u=d(n),m=a,h=u["".concat(s,".").concat(m)]||u[m]||c[m]||o;return n?r.createElement(h,i(i({ref:t},p),{},{components:n})):r.createElement(h,i({ref:t},p))}));function m(e,t){var n=arguments,a=t&&t.mdxType;if("string"==typeof e||a){var o=n.length,i=new Array(o);i[0]=u;var l={};for(var s in t)hasOwnProperty.call(t,s)&&(l[s]=t[s]);l.originalType=e,l.mdxType="string"==typeof e?e:a,i[1]=l;for(var d=2;d<o;d++)i[d]=n[d];return r.createElement.apply(null,i)}return r.createElement.apply(null,n)}u.displayName="MDXCreateElement"},354031:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>d,contentTitle:()=>l,default:()=>u,frontMatter:()=>i,metadata:()=>s,toc:()=>p});n(667294);var r=n(603905);function a(){return a=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e},a.apply(this,arguments)}function o(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}const i={id:"error-handling",title:"Gestion des erreurs"},l=void 0,s={unversionedId:"Concepts/error-handling",id:"version-20-R5/Concepts/error-handling",title:"Gestion des erreurs",description:"Le traitement des erreurs consiste \xe0 anticiper les erreurs pouvant survenir dans votre application et \xe0 y r\xe9pondre. 4D fournit un support complet pour la d\xe9tection et la signalisation des erreurs lors de l'ex\xe9cution, ainsi que pour l'analyse de leurs conditions.",source:"@site/i18n/fr/docusaurus-plugin-content-docs/version-20-R5/Concepts/error-handling.md",sourceDirName:"Concepts",slug:"/Concepts/error-handling",permalink:"/docs/fr/Concepts/error-handling",draft:!1,editUrl:"https://github.com/4d/docs/issues/new?title=Comment%20on%20Concepts%2Ferror-handling.md%20(20-R5)&body=Please%20enter%20your%20comment%3A",tags:[],version:"20-R5",frontMatter:{id:"error-handling",title:"Gestion des erreurs"},sidebar:"docs",previous:{title:"Conditions et boucles",permalink:"/docs/fr/Concepts/control-flow"},next:{title:"Modes interpr\xe9t\xe9 et compil\xe9",permalink:"/docs/fr/Concepts/interpreted-compiled"}},d={},p=[{value:"Erreur ou statut",id:"erreur-ou-statut",level:2},{value:"Installer une m\xe9thode de gestion des erreurs",id:"installer-une-m\xe9thode-de-gestion-des-erreurs",level:2},{value:"Port\xe9e et composants",id:"port\xe9e-et-composants",level:3},{value:"G\xe9rer les erreurs dans une m\xe9thode",id:"g\xe9rer-les-erreurs-dans-une-m\xe9thode",level:3},{value:"Exemple",id:"exemple",level:4},{value:"Utiliser une m\xe9thode de gestion des erreurs vide",id:"utiliser-une-m\xe9thode-de-gestion-des-erreurs-vide",level:3},{value:"Try(expression)",id:"tryexpression",level:2},{value:"Exemples",id:"exemples",level:3},{value:"Try...Catch...End try",id:"trycatchend-try",level:2},{value:"Exemple",id:"exemple-1",level:3}],c={toc:p};function u(e){var{components:t}=e,i=o(e,["components"]);return(0,r.kt)("wrapper",a({},c,i,{components:t,mdxType:"MDXLayout"}),(0,r.kt)("p",null,"Le traitement des erreurs consiste \xe0 anticiper les erreurs pouvant survenir dans votre application et \xe0 y r\xe9pondre. 4D fournit un support complet pour la d\xe9tection et la signalisation des erreurs lors de l'ex\xe9cution, ainsi que pour l'analyse de leurs conditions."),(0,r.kt)("p",null,"La gestion des erreurs r\xe9pond \xe0 deux besoins principaux :"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"rechercher et corriger les \xe9ventuels bugs et erreurs dans votre code pendant la phase de d\xe9veloppement,"),(0,r.kt)("li",{parentName:"ul"},"catching and recovering from unexpected errors in deployed applications; in particular, you can replace system error dialogs (disk full, missing file, etc.) with you own interface.")),(0,r.kt)("p",null,"Basically, there are two ways to handle errors in 4D. Vous pouvez :"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("a",a({parentName:"li"},{href:"#installing-an-error-handling-method"}),"install an error-handling method"),", or"),(0,r.kt)("li",{parentName:"ul"},"use a ",(0,r.kt)("a",a({parentName:"li"},{href:"#tryexpression"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try()")," keyword")," or a ",(0,r.kt)("a",a({parentName:"li"},{href:"#trycatchend-try"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure")," before pieces of code that call a function, method, or expression that can throw an error.")),(0,r.kt)("admonition",a({},{title:"Good practice",type:"tip"}),(0,r.kt)("p",{parentName:"admonition"},"Il est fortement recommand\xe9 d'installer une m\xe9thode globale de gestion des erreurs sur 4D Server, pour tout le code s'ex\xe9cutant sur le serveur. When 4D Server is not running ",(0,r.kt)("a",a({parentName:"p"},{href:"/docs/fr/Admin/cli"}),"headless")," (i.e. launched with its ",(0,r.kt)("a",a({parentName:"p"},{href:"/docs/fr/ServerWindow/overview"}),"administration window"),"), this method would avoid unexpected dialog boxes to be displayed on the server machine. In headless mode, errors are logged in the ",(0,r.kt)("a",a({parentName:"p"},{href:"/docs/fr/Debugging/debugLogFiles#4ddebuglogtxt-standard"}),"4DDebugLog file")," for further analysis.")),(0,r.kt)("h2",a({},{id:"erreur-ou-statut"}),"Erreur ou statut"),(0,r.kt)("p",null,"Many 4D class functions, such as ",(0,r.kt)("inlineCode",{parentName:"p"},"entity.save()")," or ",(0,r.kt)("inlineCode",{parentName:"p"},"transporter.send()"),", return a ",(0,r.kt)("em",{parentName:"p"},"status")," object. Cet objet permet de stocker les erreurs \"pr\xe9visibles\" dans le contexte d'ex\xe9cution, telles qu'un mot de passe invalide, une entit\xe9 verrouill\xe9e, etc., qui ne stoppe pas l'ex\xe9cution du programme. Cette cat\xe9gorie d'erreurs peut \xeatre g\xe9r\xe9e par du code habituel."),(0,r.kt)("p",null,'D\'autres erreurs "impr\xe9visibles" peuvent inclure une erreur en \xe9criture sur le disque, une panne de r\xe9seau ou toute interruption inattendue. This category of errors generates exceptions and needs to be handled through an error-handling method or a ',(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," keyword."),(0,r.kt)("h2",a({},{id:"installer-une-m\xe9thode-de-gestion-des-erreurs"}),"Installer une m\xe9thode de gestion des erreurs"),(0,r.kt)("p",null,"In 4D, all errors can be caught and handled by specific project methods, named ",(0,r.kt)("strong",{parentName:"p"},"error-handling")," (or ",(0,r.kt)("strong",{parentName:"p"},"error-catching"),") methods."),(0,r.kt)("p",null,"Une fois install\xe9s, les gestionnaires d'erreurs sont automatiquement appel\xe9s en mode interpr\xe9t\xe9 ou compil\xe9 en cas d'erreur dans l'application 4D ou l'un de ses composants. Un gestionnaire d'erreur diff\xe9rent peut \xeatre appel\xe9 en fonction du contexte d'ex\xe9cution (voir ci-dessous)."),(0,r.kt)("p",null,"To ",(0,r.kt)("em",{parentName:"p"},"install")," an error-handling project method, you just need to call the ",(0,r.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv19/help/command/en/page155.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"ON ERR CALL"))," command with the project method name and (optionnally) scope as parameters. Par exemple :"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_Errors";ek local) //Installe une m\xe9thode locale de gestion des erreurs\n')),(0,r.kt)("p",null,"To stop catching errors for an execution context and give back hand, call ",(0,r.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," with an empty string:"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("";ek local) //rend le contr\xf4le au process local\n')),(0,r.kt)("p",null,"The  ",(0,r.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page704.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Method called on error"))," command allows you to know the name of the method installed by ",(0,r.kt)("inlineCode",{parentName:"p"},"ON ERR CALL")," for the current process. Cela est particuli\xe8rement utile dans le contexte du code g\xe9n\xe9rique car il vous permet de modifier temporairement puis de restaurer la m\xe9thode de capture d'erreur :"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),' $methCurrent:=Method called on error(ek local)\n ON ERR CALL("NewMethod";ek local)\n  //If the document cannot be opened, an error is generated\n $ref:=Open document("MyDocument")\n  //Reinstallation of previous method\n ON ERR CALL($methCurrent;ek local)\n\n')),(0,r.kt)("h3",a({},{id:"port\xe9e-et-composants"}),"Port\xe9e et composants"),(0,r.kt)("p",null,"Une m\xe9thode de gestion des erreurs peut \xeatre d\xe9finie pour diff\xe9rents contextes d'ex\xe9cution :"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"for the ",(0,r.kt)("strong",{parentName:"li"},"current process"),"- a local error handler will be only called for errors that occurred in the current process of the current project,"),(0,r.kt)("li",{parentName:"ul"},"for the ",(0,r.kt)("strong",{parentName:"li"},"whole application"),"- a global error handler will be called for all errors that occurred in the application execution context of the current project,"),(0,r.kt)("li",{parentName:"ul"},"from the ",(0,r.kt)("strong",{parentName:"li"},"components"),"- this error handler is defined in a host project and will be called for all errors that occurred in the components when they were not already caught by a component handler.")),(0,r.kt)("p",null,"Exemples :"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("IO_Errors";ek local) //Installe une m\xe9thode de gestion des erreurs locale\nON ERR CALL("globalHandler";ek global) //Installe une m\xe9thode de gestion des erreurs globale\nON ERR CALL("componentHandler";ek erreurs des composants) //Installe une m\xe9thode de gestion des erreurs pour les composants\n')),(0,r.kt)("p",null,"Vous pouvez installer un gestionnaire d'erreurs global qui servira de \"fallback\" et des gestionnaires d'erreurs locaux sp\xe9cifiques pour certains process. Un gestionnaire d'erreur global est \xe9galement utile sur le serveur pour \xe9viter l'affichage de dialogues d'erreur sur le serveur lorsqu'il est ex\xe9cut\xe9 avec interface."),(0,r.kt)("p",null,"Vous pouvez d\xe9finir une seule m\xe9thode de gestion des erreurs pour l'ensemble de l'application ou diff\xe9rentes m\xe9thodes par module d'application. Toutefois, une seule m\xe9thode peut \xeatre install\xe9e par contexte d'ex\xe9cution et par projet."),(0,r.kt)("p",null,"Lorsqu'une erreur se produit, une seule m\xe9thode est appel\xe9e, comme le montre le sch\xe9ma suivant :"),(0,r.kt)("p",null,(0,r.kt)("img",{alt:"error management",src:n(385655).Z,width:"1195",height:"578"})),(0,r.kt)("h3",a({},{id:"g\xe9rer-les-erreurs-dans-une-m\xe9thode"}),"G\xe9rer les erreurs dans une m\xe9thode"),(0,r.kt)("p",null,"Dans une m\xe9thode de gestion d'erreur personnalis\xe9e, vous avez acc\xe8s \xe0 plusieurs \xe9l\xe9ments d'information qui vous aideront \xe0 identifier l'erreur :"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("p",{parentName:"li"},"des variables syst\xe8me d\xe9di\xe9es :"),(0,r.kt)("ul",{parentName:"li"},(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error")," (longint): error code"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error method")," (text): name of the method that triggered the error"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error line")," (longint): line number in the method that triggered the error"),(0,r.kt)("li",{parentName:"ul"},(0,r.kt)("inlineCode",{parentName:"li"},"Error formula")," (text): formula of the 4D code (raw text) which is at the origin of the error.")))),(0,r.kt)("admonition",a({},{type:"info"}),(0,r.kt)("p",{parentName:"admonition"},"4D automatically maintains a number of variables called ",(0,r.kt)("strong",{parentName:"p"},"system variables"),", meeting different needs. Consultez le manuel Language de 4D","*",".")),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"the ",(0,r.kt)("a",a({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Last errors"))," command that returns a collection of the current stack of errors that occurred in the 4D application. You can also use the ",(0,r.kt)("a",a({parentName:"li"},{href:"https://doc.4d.com/4dv19/help/command/en/page1015.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"GET LAST ERROR STACK"))," command that returns the same information as arrays."),(0,r.kt)("li",{parentName:"ul"},"the ",(0,r.kt)("inlineCode",{parentName:"li"},"Get call chain")," command that returns a collection of objects describing each step of the method call chain within the current process.")),(0,r.kt)("h4",a({},{id:"exemple"}),"Exemple"),(0,r.kt)("p",null,"Voici un syst\xe8me de gestion des erreurs simple :"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'//installer la m\xe9thode de gestion d\'erreur\nON ERR CALL("errorMethod")\n //... ex\xe9cuter le code\n ON ERR CALL("") //redonner le contr\xf4le \xe0 4D\n')),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'// errorMethod project method\n If(Error#1006) //this is not a user interruption\n    ALERT("The error "+String(Error)+" occurred". The code in question is: \\""+Error formula+"\\"")\n End if\n')),(0,r.kt)("h3",a({},{id:"utiliser-une-m\xe9thode-de-gestion-des-erreurs-vide"}),"Utiliser une m\xe9thode de gestion des erreurs vide"),(0,r.kt)("p",null,"Si vous souhaitez cacher la boite de dialogue d'erreur standard, vous pouvez installer une m\xe9thode de gestion d'erreurs vide. The ",(0,r.kt)("inlineCode",{parentName:"p"},"Error")," system variable can be tested in any method, i.e. outside of the error-handling method:"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'ON ERR CALL("emptyMethod") //emptyMethod exists but is empty\n$doc:=Open document( "myFile.txt")\nIf (Error=-43)\n    ALERT("File not found.")\nEnd if\nON ERR CALL("")\n')),(0,r.kt)("h2",a({},{id:"tryexpression"}),"Try(expression)"),(0,r.kt)("p",null,"The ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," statement allows you to test a single-line expression in its actual execution context (including, in particular, local variable values) and to intercept errors it throws so that the 4D error dialog box is not displayed. Using ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," provides an easy way to handle simple error cases with a very low number of code lines, and without requiring an error-handling method."),(0,r.kt)("admonition",a({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If you want to try a more complex code than a single-line expression, you might consider using a ",(0,r.kt)("a",a({parentName:"p"},{href:"#trycatchend-try"}),(0,r.kt)("inlineCode",{parentName:"a"},"Try/Catch")," structure"),".")),(0,r.kt)("p",null,"The formal syntax of the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," statement is:"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"\nTry (expression) : any | Undefined\n\n")),(0,r.kt)("p",null,(0,r.kt)("em",{parentName:"p"},"expression")," can be any valid expression."),(0,r.kt)("p",null,"If an error occurred during its execution, it is intercepted and no error dialog is displayed, whether an ",(0,r.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," was installed or not before the call to ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()"),". If ",(0,r.kt)("em",{parentName:"p"},"expression")," returns a value, ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," returns the last evaluated value, otherwise it returns ",(0,r.kt)("inlineCode",{parentName:"p"},"Undefined"),"."),(0,r.kt)("p",null,"You can handle the error(s) using the ",(0,r.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Last errors"))," command. If ",(0,r.kt)("em",{parentName:"p"},"expression")," throws an error within a stack of ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," calls, the execution flow stops and returns to the latest executed ",(0,r.kt)("inlineCode",{parentName:"p"},"Try()")," (the first found back in the call stack)."),(0,r.kt)("admonition",a({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If an ",(0,r.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," is installed by ",(0,r.kt)("em",{parentName:"p"},"expression"),", it is called in case of error.")),(0,r.kt)("h3",a({},{id:"exemples"}),"Exemples"),(0,r.kt)("ol",null,(0,r.kt)("li",{parentName:"ol"},"You want to display the contents of a file if the file can be open without error, and if its contents can be read. Vous pouvez \xe9crire :")),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'var $text : Text\nvar $file : 4D.File := File("/RESOURCES/myFile.txt")\nvar $fileHandle : 4D.FileHandle := Try($file.open())\nIf ($fileHandle # Null)\n  $text:=Try($fileHandle.readText()) || "Error reading the file"\nEnd if\n')),(0,r.kt)("ol",a({},{start:2}),(0,r.kt)("li",{parentName:"ol"},"You want to handle the divide by zero error. In this case, you want to return 0 and throw an error:")),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'function divide( $p1: real; $p2: real)-> $result: real\n  if ($p2=0)\n     $result:=0 //only for clarity (0 is the default for reals)\n     throw(-12345; "Division by zero")\n  else\n    $result:=$p1/$p2\n  end if\n\nfunction test()\n  $result:=Try(divide($p1;$p2))\n  If (Last errors # null)\n    ALERT("Error")\n  End if\n\n')),(0,r.kt)("ol",a({},{start:3}),(0,r.kt)("li",{parentName:"ol"},"You want to handle both ",(0,r.kt)("a",a({parentName:"li"},{href:"#error-or-status"}),"predictable and non-predictable")," errors:")),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),'var $e:=ds.Employee.new()\n$e.name:="Smith"\n$status:=Try($e.save()) //catch predictable and non-predictable errors\nIf ($status.success)\n   ALERT( "Success")\nElse\n   ALERT( "Error: "+JSON Stringify($status.errors))\nEnd if\n\n')),(0,r.kt)("h2",a({},{id:"trycatchend-try"}),"Try...Catch...End try"),(0,r.kt)("p",null,"The ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure allows you to test a block code in its actual execution context (including, in particular, local variable values) and to intercept errors it throws so that the 4D error dialog box is not displayed."),(0,r.kt)("p",null,"Unlike the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try(expression)")," keyword that evaluates a single-line expression, the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure allows you to evaluate any code block, from the most simple to the most complex, without requiring an error-handling method. In addition, the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," block can be used to handle the error in any custom way."),(0,r.kt)("p",null,"The formal syntax of the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structure is:"),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"\nTry \n    statement(s) // Code to evaluate\nCatch\n    statement(s) // Code to execute in case of error\nEnd try\n\n")),(0,r.kt)("p",null,"The code placed between the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," and the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," keywords is first executed, then the flow depends on the error(s) encountered during this execution."),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"If no error is thrown, the code execution continues after the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"End try")," keyword. The code placed between the ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," and the ",(0,r.kt)("inlineCode",{parentName:"li"},"End try")," keywords is not executed."),(0,r.kt)("li",{parentName:"ul"},"If the code block execution throws a ",(0,r.kt)("em",{parentName:"li"},"non-deferred error"),", the execution flow stops and executes the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," code block."),(0,r.kt)("li",{parentName:"ul"},"If the code block execution throws a ",(0,r.kt)("em",{parentName:"li"},"deferred error"),", the execution flow continues until the end of the ",(0,r.kt)("inlineCode",{parentName:"li"},"Try")," block and then executes the corresponding ",(0,r.kt)("inlineCode",{parentName:"li"},"Catch")," code block.")),(0,r.kt)("admonition",a({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If a ",(0,r.kt)("em",{parentName:"p"},"deferred")," error is thrown outside of the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," block, the code execution continues until the end of the method or function.")),(0,r.kt)("admonition",a({},{type:"info"}),(0,r.kt)("p",{parentName:"admonition"},"For more information on ",(0,r.kt)("em",{parentName:"p"},"deferred")," and ",(0,r.kt)("em",{parentName:"p"},"non-deferred")," errors, please refer to the ",(0,r.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20R/help/command/en/page1805.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"throw"))," command description.")),(0,r.kt)("p",null,"In the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," code block, you can handle the error(s) using standard error handling commands. The ",(0,r.kt)("a",a({parentName:"p"},{href:"https://doc.4d.com/4dv20/help/command/en/page1799.html"}),(0,r.kt)("inlineCode",{parentName:"a"},"Last errors"))," function contains the last errors collection. You can ",(0,r.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"declare an error-handling method")," in this code block, in which case it is called in case of error (otherwise the 4D error dialog box is displayed)."),(0,r.kt)("admonition",a({},{type:"note"}),(0,r.kt)("p",{parentName:"admonition"},"If an ",(0,r.kt)("a",a({parentName:"p"},{href:"#installing-an-error-handling-method"}),"error-handling method")," is installed in the code placed between the ",(0,r.kt)("inlineCode",{parentName:"p"},"Try")," and the ",(0,r.kt)("inlineCode",{parentName:"p"},"Catch")," keywords, it is called in case of error.")),(0,r.kt)("h3",a({},{id:"exemple-1"}),"Exemple"),(0,r.kt)("p",null,"Combining transactions and ",(0,r.kt)("inlineCode",{parentName:"p"},"Try...Catch...End try")," structures allows writing secured code for critical features."),(0,r.kt)("pre",null,(0,r.kt)("code",a({parentName:"pre"},{className:"language-4d"}),"Function createInvoice($customer : cs.customerEntity; $items : Collection; $invoiceRef : Text) : cs.invoiceEntity\n    var $newInvoice : cs.invoiceEntity\n    var $newInvoiceLine : cs.invoiceLineEntity\n    var $item : Object\n    ds.startTransaction()\n    Try\n        $newInvoice:=This.new()\n        $newInvoice.customer:=$customer\n        $newInvoice.invoiceRef:=$invoiceRef\n        For each ($item; $items)\n            $newInvoiceLine:=ds.invoiceLine.new()\n            $newInvoiceLine.item:=$item.item\n            $newInvoiceLine.amount:=$item.amount\n            $newInvoiceLine.invoice:=$newInvoice\n            //call other specific functions to validate invoiceline\n            $newInvoiceLine.save()\n        End for each \n        $newInvoice.save()\n        ds.validateTransaction()\n    Catch\n        ds.cancelTransaction()\n        ds.logErrors(Last errors)\n        $newInvoice:=Null\n    End try\n    return $newInvoice\n\n")))}u.isMDXComponent=!0},385655:(e,t,n)=>{n.d(t,{Z:()=>r});const r=n.p+"assets/images/error-schema-973d40c4618e79de87f884622e4e930f.png"}}]);